MEMMODEL_CONSTRAINTS.md

```diff
@@ -103,6 +103,19 @@ may perceive in that statement.)
   for memory used for atomics.  The situation is somewhat better in
   well-typed C++ but reality is complicated.
 
+## Implementation targeting C and C++ or LLVM
+
+* _Targeting C or C++_: it should be possible to implement
+  SharedArrayBuffer using C or C++.  The C/C++11 memory model allows
+  accesses with a number of orderings, and there should be a mapping
+  from SharedArrayBuffer accesses to C/C++11 orderings. Racy
+  SharedArrayBuffer accesses are required to produce defined values,
+  so cannot be implemented using C/C++ non-atomics.
+* _Targeting LLVM_: implementations may also decide to target
+  LLVM, which may provide for more optimization than targeting C/C++
+  and compiling to LLVM. In particular, LLVM allows unordered access,
+  which does not require per-variable sequential consistency.
```

Comment:
> If we map SAB non-atomic accesses to LLVM's unordered, I think this all the same optimizations as LLVM non-atomic. Slightly annoyingly, C11's relaxed blocks some optimizations due to coherence.

`unordered` isn't the same as LLVM's non-atomics.

`relaxed` basically only enforces same-address writes are seen in a thread's write order. You can totally reorder them, fuse them, move them around, eliminate dead ones, etc.

> * _Targeting C or C++_: it should be possible to implement
>  +  SharedArrayBuffer using C or C++.  The C/C++11 memory model allows

The wording is confusing: C/C++ should be mappable to SAB. Or are you saying that the implementation of SAB (e.g. a JS engine) is in C++? That's not really true either, because the SAB code is JIT-compiled.

>  +  accesses with a number of orderings, and there should be a mapping
>  +  from SharedArrayBuffer accesses to C/C++11 orderings. Racy

That's currently not true: only `seq_cst` is supported by SAB. There are issues open for other ones.

>  +  SharedArrayBuffer accesses are required to produce defined values,
>  +  so cannot be implemented using C/C++ non-atomics.

That's not really accurate: they simply don't have the same semantics. C++ has much looser guarantees, both when it comes to races (they're simply UB) and alignment. As above I don't see SAB as being "implemented using C++".

Critically for SAB:
* Most C++ code compiled to JS / wasm will already have been optimized so SAB's spec can nail down some behavior without too much cost.
* SAB doesn't try to support as many odd ISAs as C++ does, making its specification easier.

>  +* _Targeting LLVM_: implementations may also decide to target
>  +  the LLVM memory model, for example by using a language such as
>  +  Rust, which inherits its memory model from LLVM. LLVM has a very
>  +  similar model to C/C++, but allows mixed-size access, and
>  +  allows Java-like unordered access.

I still don't think this is accurate. Implementations of what? Are we talking about Rust -> LLVM -> SAB, or about ??? -> SAB -> LLVM?

I'd also like to clarify what you mean by Rust. Sure it uses LLVM, but I don't think it's accurate to say that its memory model is LLVM's.